### PR TITLE
Fix TCP fallback with multiple nameservers [Bug #8285]

### PR DESCRIPTION
**Pull request by opti**
Originally reported here https://bugs.ruby-lang.org/issues/8285, but the patch has never been finalized.

### We stumbled on the same issue under the following conditions:
- Multiple nameservers are configured for Resolv::DNS
- A nameserver falls back from UDP to TCP requester due to the message size
- TCP request hits `Resolv::DNS` timeout
- `Resolv::DNS` retries the next nameserver

#### Expected result:
`Resolv::DNS` successfully falls back to the next nameserver, using TCP request.

#### Actual result:
`Resolv::DNS::Requester::RequestError: host/port don't match` is raised.

### This pull request:
- Finalizes [original patch](https://bugs.ruby-lang.org/attachments/3665) proposed by @jmehnle.
- Reuses open TCP connection (by @hanazuki)
- Adds a test case for the scenario above.

### CODE REVIEW

**Comment by opti** [comment]
@sorah any chance this can be reviewed?

**Comment by opti** [comment]
@nobu any chance this can be reviewed?

**Review by hanazuki — COMMENTED** [review]
The problem is that the TCP requester is connected to the nameserver's address and port and cannot be used to send requests to the other nameservers. This patch solves the problem by making a TCP requester for each nameserver and memoizing it. I think this is the correct approach. (I just noticed that the memoization key should include the nameserver's port, too).

This patch also tracks the necessity of TCP fallback for each candidate FQDN and always tries UDP before TCP.  Unlike 11 years ago when the original patch was written and we had many unreliable TCP nameservers and middleboxes, the use of TCP is recommended today when possible without additional handshake overhead. I think it's better to follow this guideline:

[RFC 7766 §5](https://datatracker.ietf.org/doc/html/rfc7766#section-5)
> Stub resolvers and recursive resolvers MAY elect to send either TCP or UDP queries depending on local operational reasons.
> TCP MAY be used before sending any UDP queries.
> If the resolver already has an open TCP connection to the server, it SHOULD reuse this connection.

Another drawback of this patch is that it makes a UDP requester for each nameserver. A UDP requester is unconnected when multiple nameservers are configured (`Requester::UnconnectedUDP`), so we only need a single instance of it. The patched version creates up to (2 × #nameservers) UDP sockets when both IPv4 and IPv6 resolvers are configured. I think this needs fixing before merging the patch.

My proposal: https://github.com/ruby/resolv/compare/master...hanazuki:ruby-resolv:reuse-tcp-conn?expand=1

**Comment by opti** [comment]
@hanazuki These are valid points. Thank you for the detailed explanation. I've added your commit to this pull request. Thanks for looking into it.

**Comment by opti** [comment]
@hanazuki any chance this fix can be merged?

**Comment by hanazuki** [comment]
+1 to merge. I'm not the maintainer and cannot merge or approve PRs.

What do you think, @sorah @nobu?

**Comment by opti** [comment]
Any thoughts ☝️ @sorah, @nobu, @hsbt?

**Review by sorah — APPROVED** [review]

**Pull request merged**
